File: cses.fi/BuildingTeams.java
/*
 * https://cses.fi/problemset/task/1668/
 */
import java.io.*;
import java.util.*;

// Getting TLE, but correct approach! :(
public class BuildingTeams {
	public static void main(String[] args) throws IOException {
		BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(System.in));
		String[] inputLine = bufferedReader.readLine().trim().split(" ");
		int nodes = Integer.parseInt(inputLine[0]);
		int edges = Integer.parseInt(inputLine[1]);
		Map<Integer, List<Integer> > graph = new HashMap<>();
		while (edges-- > 0) {
			inputLine = bufferedReader.readLine().trim().split(" ");
			int u = Integer.parseInt(inputLine[0]);
			int v = Integer.parseInt(inputLine[1]);
			graph.computeIfAbsent(u, node -> new ArrayList<>()).add(v);
			graph.computeIfAbsent(v, node -> new ArrayList<>()).add(u);
		}
		int[] nodeColour = new int[nodes + 1];
		boolean isBipartitePossible = true;
		for (int node = 1; node <= nodes; node++) {
			if (nodeColour[node] != 0) {
				continue;
			}
			nodeColour[node] = 1;
			Queue<Integer> q = new LinkedList<>();
			q.offer(node);
			while (!q.isEmpty()) {
				int vertex = q.poll();
				List<Integer> neighbors = graph.getOrDefault(vertex, Collections.emptyList());
				for (int neighbor: neighbors) {
					if (nodeColour[neighbor] == nodeColour[vertex]) {
						isBipartitePossible = false;
						break;
					}
					if (nodeColour[neighbor] == 0) {
						nodeColour[neighbor] = 3 - nodeColour[vertex];
						q.offer(neighbor);
					}
				}
			}
			if (!isBipartitePossible) {
				break;
			}
		}
		if (!isBipartitePossible) {
			System.out.println("IMPOSSIBLE");
		} else {
			for (int i = 1; i <= nodes; i++) {
				System.out.print(nodeColour[i] + " ");
			}
		}
	}
}

File: cses.fi/BuildingTeams.cpp
/*
 * https://cses.fi/problemset/task/1668/
 */
#include <bits/stdc++.h>
using namespace std;

int main() {
	int nodes, edges, u, v;
	vector<vector<int> > graph;
	cin >> nodes >> edges;
	graph.resize(nodes + 1);
	while (edges--) {
		cin >> u >> v;
		graph[u].push_back(v);
		graph[v].push_back(u);
	}
	vector<int> colour(nodes + 1);
	bool isBipartite = true;
	for (int node = 1; node <= nodes; node++) {
		if (colour[node]) {
			continue;
		}
		colour[node] = 1;
		queue<int> q;
		q.push(node);
		while (!q.empty()) {
			int vertex = q.front();
			q.pop();
			for (int neighbor: graph[vertex]) {
				if (colour[neighbor] == colour[vertex]) {
					isBipartite = false;
					break;
				}
				if (!colour[neighbor]) {
					colour[neighbor] = 3 - colour[vertex];
					q.push(neighbor);
				}
			}
			if (!isBipartite) {
				break;
			}
		}
		if (!isBipartite) {
			break;
		}
	}
	if (!isBipartite) {
		cout << "IMPOSSIBLE";
	} else {
		for (int i = 1; i <= nodes; i++) {
			cout << colour[i] << ' ';
		}
	}
}
